### ISSUE TITLE
Copy files between file systems

### ISSUE THREAD

**Issue by phillipsnick** [KnpLabs/Gaufrette#346]
Is there any simple way to do this currently? Can't seem to find anything.

**Comment by NiR-** [comment]
Hey @phillipsnick, 

Sorry for the late response.

You can use streams and native `copy` function to do this. I've just added a functional test case showing how to do [here](https://github.com/KnpLabs/Gaufrette/pull/367/files). You can do the same to copy files across multiple filesystems.

Hope it helps you, let me know if you encounter some problems.

**Comment by phillipsnick** [comment]
No worries, I ended up putting something similar together in the end. Cheers :+1:

**Comment by SrgSteak** [comment]
Does this work with large files too? For example, upload a 5GB file from a local storage to dropbox?

**Comment by akovalyov** [comment]
@SrgSteak we haven't tested it with large files, but I don't see the reason why Gaufrette should be the main trouble for that, since it is just native `copy` PHP function .
 More troubles can be occured by:
* PHP timeouts
* memory limits
* network connection issues

**Comment by SrgSteak** [comment]
@akovalyov 
PHP timeouts and network connections shouldn't be a problem (background worker). Memory limits is another story. Any recommendations to keep memory usage low?

**Comment by akovalyov** [comment]
Under memory usage limits I meant the amount of memory which will actually be consumed by `copy` function. But according to [that awesome SO comment](http://stackoverflow.com/a/18497907), it won't consume too much. So, everything should work fine.
The only thing I want to mention is FS limitations, but if you are not copying the files to FAT32|NTFS|etc, you shouldn't have any troubles there. 

However this discussion is not really related to Gaufrette, I kindly ask you to write a comment about your results when you try it so that we can have some real examples of this particular usecase. Does it work for you?

**Comment by SrgSteak** [comment]
@akovalyov I will tell you as soon as i get my AWS connection to work |D

**Comment by raistlin** [comment]
Where can I find what adapters support streaming?

2016-02-04 15:43 GMT+01:00 rs <notifications@github.com>:

> @akovalyov <https://github.com/akovalyov> I will tell you as soon as i
> get my AWS connection to work |D
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/KnpLabs/Gaufrette/issues/346#issuecomment-179876470>.
>



-- 
Raúl Ferriz
Ingeniero Técnico en Informática

**Comment by SrgSteak** [comment]
@akovalyov So, i finally got aws_s3 filesystem to work. The documentation doesn't work with version 3.x of aws/aws-sdk-php. (http://stackoverflow.com/a/31582475)
I successfully uploaded a 250MB video to a bucket, but only after increasing the php memory. I run into a fatal error (PHP Fatal error:  Allowed memory size of ... exhausted in vendor/knplabs/gaufrette/src/Gaufrette/Stream/InMemoryBuffer.php on line 90) with bigger files.

**Comment by akovalyov** [comment]
@SrgSteak OK thank you for the report! Could you open a new issue for that so that we can track this bug? This issue is not very relevant :(

**Comment by SrgSteak** [comment]
@akovalyov a bug for the memory exhaustion or the aws_s3 configuration?

**Comment by akovalyov** [comment]
@SrgSteak brilliant question. I'd say, if you have several minutes to open two separate issues for that - would be perfect. As for memory limit, could you also add your current memory limit there? 256 MB, I assume?

**Comment by SrgSteak** [comment]
@akovalyov exactly. I started with the default of 128MB for the upload of the 250MB leading to the memory problem -> increasing the limit to 256MB worked, but bigger files lead to the same error. I'll open 2 issues.

**Comment by SrgSteak** [comment]
Issues are open now.
AWS config: #376
Memory problems: #377

I'd like to help solving those issues, but honestly, i'm not sure where to start.

**Comment by akovalyov** [comment]
@SrgSteak thank you, let's continue the conversation in the corresponding issues, What I definitely can tell now is that AWS has higher priority for us right now and as for memory leaks, I am not sure yet if we are able to configure it. It will require the rewriting InMemoryBuffer or new ChunkedMemoryBuffer to upload files larger then available memory for php process.
